src/core/symbology/qgssymbollayerutils.cpp

```diff
@@ -2947,33 +2949,49 @@ QgsStringMap QgsSymbolLayerUtils::getVendorOptionList( QDomElement &element )
 }
 
 
-QgsStringMap QgsSymbolLayerUtils::parseProperties( QDomElement &element )
+QVariantMap QgsSymbolLayerUtils::parseProperties( const QDomElement &element )
 {
-  QgsStringMap props;
-  QDomElement e = element.firstChildElement();
-  while ( !e.isNull() )
+  QVariant newSymbols = QgsXmlUtils::readVariant( element ).toMap();
+  if ( newSymbols.type() == QVariant::Map )
+  {
+    return newSymbols.toMap();
+  }
+  else
   {
-    if ( e.tagName() == QLatin1String( "prop" ) )
+    // read old style of writing properties
+    // backward compatibility with project saved in <= 3.16
+    QVariantMap props;
+    QDomElement e = element.firstChildElement();
+    while ( !e.isNull() )
     {
-      QString propKey = e.attribute( QStringLiteral( "k" ) );
-      QString propValue = e.attribute( QStringLiteral( "v" ) );
-      props[propKey] = propValue;
+      if ( e.tagName() == QLatin1String( "prop" ) )
+      {
+        QString propKey = e.attribute( QStringLiteral( "k" ) );
+        QString propValue = e.attribute( QStringLiteral( "v" ) );
+        props[propKey] = propValue;
+      }
+      e = e.nextSiblingElement();
     }
-    e = e.nextSiblingElement();
+    return props;
   }
-  return props;
 }
 
 
-void QgsSymbolLayerUtils::saveProperties( QgsStringMap props, QDomDocument &doc, QDomElement &element )
+void QgsSymbolLayerUtils::saveProperties( QVariantMap props, QDomDocument &doc, QDomElement &element )
```

Comment:
one of the most important changes: we now use QgsXmlUtils to save the properties, which is different from the old behavior. I kept the old saving method to have a backward compatibility as much as possible at the cost of extra lines in the XML.

Reply:
That's the right move, thanks